### DISCUSSION TITLE
Write custom tags to files on import

### DISCUSSION THREAD

**Discussion by AlexChalk** [beetbox/beets#4293] [Q&A]
Hi,

My mp3 player cares about the 'AlbumSort' tag, and I'd like to add it automatically when importing files into my beets library.

I'm imagining a config that looks a lot like `paths:`:

```
custom_tags:
  AlbumSort: $year $album # Sorts albums by year, then alphabetically
```

Does this look like a viable feature, and if not can you point me in the right direction for authoring a plugin that would add an `AlbumSort` tag in this way? Thanks!

**Comment by AlexChalk** [comment]
According to https://beets.readthedocs.io/en/stable/dev/plugins.html it looks like something like this should do it:

```
from beets.plugins import BeetsPlugin
from beets import config
import mediafile


# pip install ./beets-albumsort
class AlbumSortPlugin(BeetsPlugin):
    def __init__(self, *args, **kwargs):
        super(AlbumSortPlugin, self).__init__(*args, **kwargs)
        field = mediafile.MediaField(
            mediafile.MP3DescStorageStyle(u'AlbumSort'),
            mediafile.StorageStyle(u'AlbumSort')
        )
        self.add_media_field('AlbumSort', field)

        self.register_listener('write', self.write)

    def write(self, item, path, tags):
        tags["AlbumSort"] = f"{tags['original_year']} {tags['album']}"
```

But the code above does not write the tag :/.

**Reply by sampsyo** [reply]
You're on the right track here! I'm not 100% sure what's going on, but after adding a `MediaField` with the `add_media_field` method, you should be able to populate this field. Can you try removing the `write` listener and manually setting the field with, for example, `beet modify` and `beet write`? Looking into that will debug the first part of your solution.

Then, once the new tag is working, you can worry about how to automatically set it. I'd recommend trying `database_change`, which lets you modify a field every time something gets added to, removed from, or modified in the database itself—which should then get propagated to files on write.

**Reply by AlexChalk** [reply]
Hi @sampsyo, thanks for this. `modify` + `write` doesn't populate the field, exit code is 0 but tags aren't updated. So I guess I'm declaring the MediaField incorrectly.

Maybe my `MP3DescStorageStyle` and `StorageStyle` are incorrect, the example in the docs sets them both to the tag name, but this does seem a bit weird. Do you have any idea how those args are used? Thanks!

**Reply by sampsyo** [reply]
For more examples, you can see the MediaFile source code:
https://github.com/beetbox/mediafile/blob/4f70ee8a29456ca98c071e59f08310b2b9948afb/mediafile.py#L1994-L1999

**Reply by AlexChalk** [reply]
That's great, it looks like the names are important! This seems to do the trick:
```
from beets.plugins import BeetsPlugin
from beets import config
import mediafile


# pip install ./beets-albumsort
class AlbumSortPlugin(BeetsPlugin):
    def __init__(self, *args, **kwargs):
        album_sort_field = mediafile.MediaField(
            mediafile.MP3StorageStyle(u"TSOA"),
            mediafile.MP4StorageStyle(u"----:com.apple.iTunes:Album Sort"),
            mediafile.StorageStyle(u"Album Sort")
        )
        self.add_media_field("album_sort", album_sort_field)

        self.register_listener("write", self.write)

        super().__init__(*args, **kwargs)

    def write(self, item, path, tags):
        tags["album_sort"] = f"{tags['original_year']} {tags['album']}"
```
